FOLIO: automatically renew expired token as needed.

demiankatz opened the pull request:
I ran into some problems with newer FOLIO versions encountering expired tokens during long-running processes (like rebuilding the course reserves index). This PR adds a mechanism to auto-renew expired tokens as needed.

demiankatz committed:
FOLIO: automatically renew expired token as needed.

maccabeelevine commented on module/VuFind/src/VuFind/ILS/Driver/AbstractAPI.php:
```diff
@@ -127,6 +143,8 @@ protected function failureCodeIsAllowed(int $code, $allowedFailureCodes): bool
      * expression, or boolean true to allow all codes.
      * @param string|array      $debugParams         Value to use in place of $params
      * in debug messages (useful for concealing sensitive data, etc.)
+     * @param int               $retryNumber         Counter to keep track of retries
```
Semantically I prefer `$attemptNumber` so you can start at 1 without having to explain why $retryNumber=0 is a thing.  You can also avoid the +1 in the log message.

maccabeelevine reviewed (approved):
Looks good and works.  One nitpick suggestion but not important.

maccabeelevine commented:
This also continues to ignores the whole renewal token option, but if you don't need it, why bother?

demiankatz committed:
Review feedback.

demiankatz commented:
Thanks, @maccabeelevine! I've adjusted as per your suggestion, which does make everything a bit more readable. Also added a missing "unused parameters" annotation. I'll see if I can find time to add test coverage as well before I merge this.

demiankatz committed:
Merge remote-tracking branch 'origin/dev' into folio-auto-renew

demiankatz commented:
Unfortunately, I've discovered that testing this in our existing unit test is not straightforward since the current test suite mocks out the makeRequest method. However, I've run a local course reserves reindex using the updated code (which is a long-running process that exercises the new logic) and it worked successfully, so I will go ahead and merge this. We can improve test coverage in the future as time permits.
